Historical humidity processing

Rea888 opened the pull request.

Rea888 committed:
created HistoricalHumidityProcessingReportsModel.php
created crate_historical_humidity_processing_reports_models_table.php

Rea888 committed:
created city_date.php

Rea888 committed:
created saveHumidityProcessToDb();
changed sequence of array in city_date array
created HistoricalHumidityProcessCommand

KuveeTheBird commented on app/Service/HistoricalHumidityProcessingService.php:
```diff
@@ -0,0 +1,31 @@
+            foreach (config('city_date.city_date')['month'] as $month) {
+                foreach (config('city_date.city_date')['city'] as $city) {
+                    if (intval($year) == 2023 && intval($month) >= 02) {
+                        exit();
```
with `exit()` you forcefully stop the code execution, and that's not what we want. we want the code to reach it's natural conclusion, so we should just break out of all the foreach statements, like this:
```suggestion
                        break 3;
```

KuveeTheBird replied:
suggested reading:
https://www.php.net/manual/en/control-structures.break.php

KuveeTheBird replied:
if the you make the changes requested above, this change will be unncessary

KuveeTheBird commented on app/Service/HistoricalHumidityProcessingService.php:
```diff
@@ -0,0 +1,31 @@
+        foreach (config('city_date.city_date')['year'] as $year) {
+            foreach (config('city_date.city_date')['month'] as $month) {
+                foreach (config('city_date.city_date')['city'] as $city) {
+                    if (intval($year) == 2023 && intval($month) >= 02) {
```
while this is a "working" solution, this way you would have to push new code changes every month until the day the unvicerse collapses into a black hole...

I would advise that you only store a startDate in your  config, e.g.: 2013-01-01 00:00:00

Then instead of using foreach statement for the dates, maybe calculate runtime the current date (e.g.: today it's 2023-02) and then iterate between startDate and today.
simple example:
`
$startDate = new DateTime(config('start_date'));
$today = new DateTime();
$cities = config('cities');

while ($startDate < $today) {
  foreach ($cities as $city) {
    HistoricalHumidityProcessingReportsModel::firstOrCreate(...)
  }
  $startDate = $startDate->add(new DateInterval('P1M'));
}
`

KuveeTheBird replied:
suggested reading:
https://www.phptutorial.net/php-oop/php-datetime/
https://www.php.net/manual/en/class.datetime.php
https://www.php.net/manual/en/class.dateinterval.php

KuveeTheBird commented on config/services.php:
```diff
@@ -35,4 +35,5 @@
         'key' => env('GOOGLE_KEY'),
     ],
 
+
```
remove unnecessary empty lines

Rea888 replied:
fixed the necessary changes

Rea888 committed:
Deleted wrong return value
Fixed saveHumidityProcessToDb() with DateTime instead of arrays
Deleted arrays from city_date config
Deleted empty rows from services.php

KuveeTheBird commented on app/Service/HistoricalHumidityProcessingService.php:
```diff
@@ -9,23 +9,24 @@ class HistoricalHumidityProcessingService
-                        );
-                    }
+        try {
+            $startDate = new \DateTime(implode('","', config('city_date.start_date')));
```
what is this implode here, why??????? :D

- [x] remove implode

KuveeTheBird commented on app/Service/HistoricalHumidityProcessingService.php:
```diff
@@ -9,23 +9,24 @@ class HistoricalHumidityProcessingService
+                foreach ($cities as $city) {
+                    HistoricalHumidityProcessingReportsModel::firstOrCreate(
+                        [
+                            'year' => $startDate->format('Y'),
```
nice, i hoped you figure out to use format, nice, congrats :)

KuveeTheBird commented on routes/web.php:
```diff
@@ -17,20 +17,6 @@
 });
 
-Route::get('/save', [\App\Http\Controllers\HistoricalWeatherHumidityController::class, 'saveHumidityToDb']);
+Route::get('/save', [\App\Service\HistoricalHumidityProcessingService::class, 'saveHumidityProcessToDb']);
```
you can "import" the HistoricalHumidityProcessingService, so only the class name is shown here

- [x] use/import class (simplify fqcn in PhpStorm)

KuveeTheBird commented on config/city_date.php:
```diff
@@ -0,0 +1,15 @@
+    'start_date' => [
+        '2013-01'
+    ]
+
```
there is still an unnecessary empty line :D

- [x] remove empty line

KuveeTheBird reviewed (changes requested):
only small things now :)

KuveeTheBird commented on config/city_date.php:
```diff
@@ -0,0 +1,15 @@
+        'Tokyo',
+    ],
+
+    'start_date' => [
```
```suggestion
    'start_date' => '2013-01'
```

Rea888 committed:
Deleted implode function, changed value in config file from array to string
Deleted empty lines in city_date.php config file
Imported class to web.php
